[WGSL] Implement bitwise expression parsing

djg opened the pull request:
#### dbe54de0044f94bf3d6a7fbf07fc29ace272bb58
<pre>
[WGSL] Implement bitwise expression parsing
<a href="https://bugs.webkit.org/show_bug.cgi?id=251670">https://bugs.webkit.org/show_bug.cgi?id=251670</a>
rdar://104997452

Reviewed by Tadeu Zagallo.

Handle parsing of expression involving &amp;, ^ and | according to the WGSL spec.

Canonical link: <a href="https://commits.webkit.org/260167@main">https://commits.webkit.org/260167@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/8a4bac691821682506706e7159312997eff6a84f

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/107387 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/16448 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/40265 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/116547 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/115972 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/111287 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/17871 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/7687 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/99549 "Built successfully") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/113150 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/13483 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/96642 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/41106 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/95386 "Passed tests") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/28268 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/82871 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/9458 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/29621 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/10112 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/6539 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/15614 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/49202 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/7032 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/11670 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

tadeuzagallo commented on Source/WebGPU/WGSL/Parser.cpp:
```diff
@@ -681,22 +721,9 @@ Expected<UniqueRef<AST::Expression>, Error> Parser<Lexer>::parseAdditiveExpressi
-            RELEASE_ASSERT_NOT_REACHED_WITH_MESSAGE("Expected + or -");
-        }
-
+        // parseMultiplicativeExpression handles multiplicative operators so
+        // token should be PLUS or MINUS.
```
Is it worth adding a debug assertion here?

tadeuzagallo commented on Source/WebGPU/WGSL/Lexer.cpp:
```diff
@@ -131,6 +134,12 @@ Token Lexer<T>::lex()
         return makeToken(TokenType::Plus);
+    case '^':
+        shift();
+        return makeToken(TokenType::Xor);
```
Eventually, we should refactor the tokens so that all the single char tokens have the same `TokenType` and we use the char to disambiguate. That gives us printing for free and doesn't require naming the token.

djg committed:
[WGSL] Implement bitwise expression parsing
https://bugs.webkit.org/show_bug.cgi?id=251670
rdar://104997452

Reviewed by Tadeu Zagallo.

Handle parsing of expression involving &, ^ and | according to the WGSL spec.

Canonical link: https://commits.webkit.org/260167@main
